### DISCUSSION TITLE
Support for Geo(Location based) Queries in Postgresql

### DISCUSSION THREAD

**Discussion by ketulrastogi** [serverpod/serverpod#602] [Q&A] [answered]
How can we perform geo queries to find nearby locations from Postgresql?
Is this feature currently available?

If not, is it on the roadmap for the near future?

Thanks

**Comment by vlidholt** [comment, marked as answer]
There is no built-in support in the ORM for location-based queries. This may be a nice feature to add in the future. However, you can always make raw SQL queries to access all the features of Postgres. More information on how to do that is here:
https://docs.serverpod.dev/concepts/database-communication#executing-raw-queries

**Comment by fleeser** [comment]
As @vlidholt mentioned, you could use raw SQL queries for now.
I'd like to share the basic steps to do so for other users wanting to know about how to do Geo queries.

1. Make sure that you have columns for latitude and longitude
2. Execute the following commands on your database (you can use tools like pgAdmin):
```sql
CREATE EXTENSION cube
```
```sql
CREATE EXTENSION earthdistance
```
3. CREATE your SQL query and run it on the server
```dart
String lat_column = "lat";
String long_column = "long";
int max_distance_in_meters = 10000;
double current_user_lat = 69.6969;
double current_user_long = 69.69.69;

String query = '''
    SELECT *, ROUND(earth_distance(ll_to_earth($current_user_lat, $current_user_long), ll_to_earth($lat_column, $long_column))::NUMERIC, 2) AS distance
    FROM
        events
    WHERE
        earth_box(ll_to_earth ($current_user_lat, $current_user_long), max_distance_in_meters) @> ll_to_earth ($lat_column, $long_column)
        AND earth_distance(ll_to_earth ($current_user_lat, $current_user_long), ll_to_earth ($lat_column, $long_column)) < max_distance_in_meters
    ORDER BY
        distance;'''
        
List<List<dynamic>> results = await session.db.query(query);
```

This will query all events that are less than 10000 meters away from the current users location, ordered by distance.

Based on an amazing article: https://postindustria.com/postgresql-geo-queries-made-easy/

**Comment by iamshivkumar** [comment]
PostGIS extension used for geo queries. It is default available on the cloud postgress. Just need to enable that. But PostGis support is needed on Serverpod to add data of type "geometry(Point,4326)".  Added data looks like this, "SRID=4326;POINT(-74.006 40.7128)". 

`INSERT INTO places (name, location)
VALUES ('Pune', ST_SetSRID(ST_MakePoint(18.5204, 73.8567), 4326));
`

`SELECT *
FROM places
WHERE ST_DWithin(location, ST_SetSRID(ST_MakePoint(17.6599, 75.9064), 4326), 3)
ORDER BY ST_Distance(location, ST_SetSRID(ST_MakePoint(17.6599, 75.9064), 4326));
`
This is some reference code. As it is a long query, we can go with raw queries while listing data. At least proper serialization/deserialization is needed to do other basic queries.
